src/ParallelAlgorithms/LinearSolver/Multigrid/ElementsAllocator.hpp

```diff
@@ -0,0 +1,181 @@
+    size_t multigrid_level = 0;
+    do {
+      // Store the current grid as child grid before coarsening it
+      children_refinement_levels = initial_refinement_levels;
+      initial_refinement_levels = parent_refinement_levels;
+      // Construct coarsened (parent) grid
+      if (not max_levels.has_value() or multigrid_level < *max_levels - 1) {
+        parent_refinement_levels =
+            LinearSolver::multigrid::coarsen(initial_refinement_levels);
+      }
+      // Create element IDs for all elements on this level
+      std::vector<ElementId<Dim>> element_ids{};
+      for (const auto& block : domain.blocks()) {
+        const std::vector<ElementId<Dim>> block_element_ids =
+            initial_element_ids(block.id(),
+                                initial_refinement_levels[block.id()],
+                                multigrid_level);
+        element_ids.insert(element_ids.begin(), block_element_ids.begin(),
+                           block_element_ids.end());
+      }
+      // Create an array section for this refinement level
+      std::vector<CkArrayIndex> array_indices(element_ids.size());
+      std::transform(
+          element_ids.begin(), element_ids.end(), array_indices.begin(),
+          [](const ElementId<Dim>& local_element_id) noexcept {
+            return Parallel::ArrayIndex<ElementId<Dim>>(local_element_id);
+          });
+      using MultigridLevelSection =
+          Parallel::Section<ElementArray, Tags::MultigridLevel>;
+      get<Parallel::Tags::Section<ElementArray, Tags::MultigridLevel>>(
+          initialization_items) = MultigridLevelSection{
+          multigrid_level, MultigridLevelSection::cproxy_section::ckNew(
+                               element_array.ckGetArrayID(),
+                               array_indices.data(), array_indices.size())};
+      using FinestGridSection =
+          Parallel::Section<ElementArray, Tags::IsFinestGrid>;
+      get<Parallel::Tags::Section<ElementArray, Tags::IsFinestGrid>>(
+          initialization_items) =
+          multigrid_level == 0
+              ? std::make_optional(FinestGridSection{
+                    true, FinestGridSection::cproxy_section::ckNew(
+                              element_array.ckGetArrayID(),
+                              array_indices.data(), array_indices.size())})
+              : std::nullopt;
+      // Create the elements for this refinement level and distribute them among
+      // processors
+      const int number_of_procs = sys::number_of_procs();
+      const domain::BlockZCurveProcDistribution<Dim> element_distribution{
+          static_cast<size_t>(number_of_procs), initial_refinement_levels};
+      for (const auto& element_id : element_ids) {
+        const size_t target_proc = element_distribution.get_proc_for_element(
+            element_id.block_id(), element_id);
+        element_array(element_id)
+            .insert(global_cache, initialization_items, target_proc);
+      }
+      Parallel::printf(
```

Comment:
I had decided against it because it prints only a few lines at the very start of the simulation, and it's very important information. It should be above `Verbosity::Quiet` (which prints the residual norm after every iteration).